File: examples/testqcl1.go

package main

import (
	"github.com/hrautila/go.opt/matrix"
	"github.com/hrautila/go.opt/linalg/blas"
	"github.com/hrautila/go.opt/linalg/lapack"
	"github.com/hrautila/go.opt/linalg"
	"github.com/hrautila/go.opt/cvx"
	"github.com/hrautila/go.opt/cvx/sets"
	"github.com/hrautila/go.opt/cvx/checkpnt"
	"fmt"
	"flag"
)

var xVal, zVal, AVal, bVal string
var spPath string
var maxIter int 
var spVerbose bool

// helper variables for checkpointing...
var loopg, loopf int

func init() {
	flag.BoolVar(&spVerbose, "V", false, "Savepoint verbose reporting.")
	flag.IntVar(&maxIter, "N", 30, "Max number of iterations.")
	flag.StringVar(&spPath, "sp", "", "savepoint directory")
	flag.StringVar(&xVal, "x", "", "Reference value for X")
	flag.StringVar(&zVal, "z", "", "Reference value for Z")
	flag.StringVar(&AVal, "A", "", "Problem data A")
	flag.StringVar(&bVal, "b", "", "Problem data b")
	loopg = 0
	loopf = 0
}
	
func errorToRef(ref, val *matrix.FloatMatrix) (nrm float64, diff *matrix.FloatMatrix) {
	diff = matrix.Minus(ref, val)
	nrm = blas.Nrm2(diff).Float()
	return
}

func check(x, z *matrix.FloatMatrix) {
	if len(xVal) > 0 {
		ref, _ := matrix.FloatParseSpe(xVal)
		nrm, diff := errorToRef(ref, x)
		fmt.Printf("x: nrm=%.9f\n", nrm)
		if nrm > 10e-7 {
			fmt.Printf("diff=\n%v\n", diff.ToString("%.12f"))
		}
	}
	if len(zVal) > 0 {
		ref, _ := matrix.FloatParseSpe(zVal)
		nrm, diff := errorToRef(ref, z)
		fmt.Printf("z: nrm=%.9f\n", nrm)
		if nrm > 10e-7 {
			fmt.Printf("diff=\n%v\n", diff.ToString("%.12f"))
		}
	}
}


// Internal type for MatrixG interface
type matrixFs struct {
	A *matrix.FloatMatrix
}

func (g *matrixFs)Gf(x, y *matrix.FloatMatrix, alpha, beta float64, trans linalg.Option)  error {
	//
	minor := 0
	if ! checkpnt.MinorEmpty() {
		minor = checkpnt.MinorTop()
	} else {
		loopg += 1
		minor = loopg
	}
	checkpnt.Check("00-Gfunc", minor)

	m, n := g.A.Size()
	y.Scale(beta)

	// x_n = x[:n]
	x_n := matrix.FloatVector(x.FloatArray()[:n])
	
	// x_n_2n = x[n:2*n]
	x_n_2n := matrix.FloatVector(x.FloatArray()[n:2*n])

	if linalg.Equal(trans, linalg.OptNoTrans) {
		// y += alpha * G * x

		// y[:n] += alpha * (x[:n] - x[n:2*n])
		y_n := matrix.Minus(x_n, x_n_2n).Scale(alpha)
		y.AddIndexes(matrix.Indexes(n), y_n.FloatArray())

		// y[n:2*n] += alpha * (-x[:n] - x[n:2*n]) = -alpha * (x[:n]+x[n:2*n])
		y_n = matrix.Plus(x_n, x_n_2n).Scale(-alpha)
		y.AddIndexes(matrix.Indexes(n, 2*n), y_n.FloatArray())

		// y[2*n+1:] += -alpha * A * x[:n]
		y_2n :=	matrix.Times(g.A, x_n).Scale(-alpha)
		y.AddIndexes(matrix.Indexes(2*n+1, y.NumElements()), y_2n.FloatArray())
	} else {
		// x_m = x[-m:]
		x_m := matrix.FloatVector(x.FloatArray()[x.NumElements()-m:])

		// x_tmp = (x[:n] - x[n:2*n] - A.T * x[-m:])
		x_tmp := matrix.Minus(x_n, x_n_2n, matrix.Times(g.A.Transpose(), x_m))

		// y[:n] += alpha * (x[:n] - x[n:2*n] - A.T * x[-m:])
		y.AddIndexes(matrix.Indexes(n), x_tmp.Scale(alpha).FloatArray())

		x_tmp = matrix.Plus(x_n, x_n_2n).Scale(-alpha)
		y.AddIndexes(matrix.Indexes(n, y.NumElements()), x_tmp.FloatArray())
	}
	checkpnt.Check("10-Gfunc", minor)
	return nil
}

func qcl1(A, b *matrix.FloatMatrix) (*cvx.Solution, error) {

     // Returns the solution u, z of
     //
     //   (primal)  minimize    || u ||_1       
     //             subject to  || A * u - b ||_2  <= 1
     //
     //   (dual)    maximize    b^T z - ||z||_2
     //             subject to  || A'*z ||_inf <= 1.
	 //
     // Exploits structure, assuming A is m by n with m >= n. 

	m, n := A.Size()
	Fkkt := func(W *sets.FloatMatrixSet) (f cvx.KKTFunc, err error) {

		err = nil
		f = nil
		beta := W.At("beta")[0].GetIndex(0)
		v := W.At("v")[0]

		// As = 2 * v *(v[1:].T * A)
		v_1 := matrix.FloatNew(1, v.NumElements()-1, v.FloatArray()[1:])
		As := matrix.Times(v, matrix.Times(v_1, A)).Scale(2.0)

		As_1 := As.GetSubMatrix(1, 0, m, n)
		As_1.Scale(-1.0)
		As.SetSubMatrix(1, 0, matrix.Minus(As_1, A))
		As.Scale(1.0/beta)

		S := matrix.Times(As.Transpose(), As)
		d1 := matrix.FloatVector(W.At("d")[0].FloatArray()[:n])
		d2 := matrix.FloatVector(W.At("d")[0].FloatArray()[n:])

		// D = 4.0 * (d1**2 + d2**2)**-1
		d := matrix.Plus(matrix.Mul(d1,d1), matrix.Mul(d2, d2)).Inv().Scale(4.0)
		// S[::n+1] += d
		S.AddIndexes(matrix.DiagonalIndexes(S), d.FloatArray())

		err = lapack.Potrf(S)
		if err != nil {	return }
		
		f = func(x, y, z *matrix.FloatMatrix) (err error) {
            // -- z := - W**-T * z 

			minor := 0
			if ! checkpnt.MinorEmpty() {
				minor = checkpnt.MinorTop()
			} else {
				loopf += 1
				minor = loopf
			}
			checkpnt.Check("00-f", minor)

			// z[:n] = -div( z[:n], d1 )
			z_val := matrix.FloatVector(z.FloatArray()[:n])
			z_res := matrix.Div(z_val, d1).Scale(-1.0)
			z.SetIndexes(matrix.Indexes(n), z_res.FloatArray())

			// z[n:2*n] = -div( z[n:2*n], d2 )
			z_val = matrix.FloatVector(z.FloatArray()[n:2*n])
			z_res = matrix.Div(z_val, d2).Scale(-1.0)
			z.SetIndexes(matrix.Indexes(n, 2*n), z_res.FloatArray())

			// z[2*n:] -= 2.0*v*( v[0]*z[2*n] - blas.dot(v[1:], z[2*n+1:]) ) 
			v0_z2n := v.GetIndex(0)*z.GetIndex(2*n)
			v1_z2n := blas.DotFloat(v, z, &linalg.IOpt{"offsetx", 1}, &linalg.IOpt{"offsety", 2*n+1})
			z_res = matrix.Scale(v, -2.0*(v0_z2n - v1_z2n))
			z.AddIndexes(matrix.Indexes(2*n, z.NumElements()), z_res.FloatArray())

			// z[2*n+1:] *= -1.0
			z.Scale(-1.0, matrix.Indexes(2*n+1, z.NumElements())...)

			// z[2*n:] /= beta
			z.Scale(1.0/beta, matrix.Indexes(2*n, z.NumElements())...)

			// -- x := x - G' * W**-1 * z

			// z_n = z[:n], z_2n = z[n:2*n], z_m = z[-(m+1):], 
			z_n := matrix.FloatVector(z.FloatArray()[:n])
			z_2n := matrix.FloatVector(z.FloatArray()[n:2*n])
			z_m := matrix.FloatVector(z.FloatArray()[z.NumElements()-(m+1):])

			// x[:n] -= div(z[:n], d1) - div(z[n:2*n], d2) + As.T * z[-(m+1):]
			z_res = matrix.Minus(matrix.Div(z_n, d1), matrix.Div(z_2n, d2))
			a_res := matrix.Times(As.Transpose(), z_m)
			z_res.Plus(a_res).Scale(-1.0)
			x.AddIndexes(matrix.Indexes(n), z_res.FloatArray())

			// x[n:] += div(z[:n], d1) + div(z[n:2*n], d2) 
			z_res = matrix.Plus(matrix.Div(z_n, d1), matrix.Div(z_2n, d2))
			x.AddIndexes(matrix.Indexes(n, x.NumElements()), z_res.FloatArray())
			checkpnt.Check("15-f", minor)
			
            // Solve for x[:n]:
            //
            //    S*x[:n] = x[:n] - (W1**2 - W2**2)(W1**2 + W2**2)^-1 * x[n:]

			// w1 = (d1**2 - d2**2), w2 = (d1**2 + d2**2)
            w1 := matrix.Minus(matrix.Mul(d1, d1), matrix.Mul(d2, d2))
            w2 := matrix.Plus(matrix.Mul(d1, d1), matrix.Mul(d2, d2))

			// x[:n] += -mul( div(w1, w2), x[n:])
			x_n := matrix.FloatVector(x.FloatArray()[n:])
			x_val := matrix.Mul(matrix.Div(w1, w2), x_n).Scale(-1.0)
			x.AddIndexes(matrix.Indexes(n), x_val.FloatArray())
			checkpnt.Check("25-f", minor)

            // Solve for x[n:]:
            //
            //    (d1**-2 + d2**-2) * x[n:] = x[n:] + (d1**-2 - d2**-2)*x[:n]

			//fmt.Printf("f potrs: x=\n%v\n", x.ToString("%.5f"))
			err = lapack.Potrs(S, x)
			if err != nil {
				fmt.Printf("Potrs error: %s\n", err)
			}
			checkpnt.Check("30-f", minor)

            // Solve for x[n:]:
            //
            //    (d1**-2 + d2**-2) * x[n:] = x[n:] + (d1**-2 - d2**-2)*x[:n]

			// w1 = (d1**-2 - d2**-2), w2 = (d1**-2 + d2**-2)
            w1 = matrix.Minus(matrix.Mul(d1, d1).Inv(), matrix.Mul(d2, d2).Inv())
            w2 = matrix.Plus(matrix.Mul(d1, d1).Inv(), matrix.Mul(d2, d2).Inv())
			x_n = matrix.FloatVector(x.FloatArray()[:n])

			// x[n:] += mul( d1**-2 - d2**-2, x[:n])
			x_val = matrix.Mul(w1, x_n)
			x.AddIndexes(matrix.Indexes(n, x.NumElements()), x_val.FloatArray())
			checkpnt.Check("35-f", minor)

			// x[n:] = div( x[n:], d1**-2 + d2**-2)
			x_n = matrix.FloatVector(x.FloatArray()[n:])
			x_val = matrix.Div(x_n, w2)
			x.SetIndexes(matrix.Indexes(n, x.NumElements()), x_val.FloatArray())
			checkpnt.Check("40-f", minor)

			// x_n = x[:n], x-2n = x[n:2*n]
			x_n = matrix.FloatVector(x.FloatArray()[:n])
			x_2n := matrix.FloatVector(x.FloatArray()[n:2*n])

			// z := z + W^-T * G*x 
			// z[:n] += div( x[:n] - x[n:2*n], d1) 
			x_val = matrix.Div(matrix.Minus(x_n, x_2n), d1)
			z.AddIndexes(matrix.Indexes(n), x_val.FloatArray())
			checkpnt.Check("44-f", minor)

			// z[n:2*n] += div( -x[:n] - x[n:2*n], d2) 
			x_val = matrix.Div(matrix.Plus(x_n, x_2n).Scale(-1.0), d2)
			z.AddIndexes(matrix.Indexes(n, 2*n), x_val.FloatArray())
			checkpnt.Check("48-f", minor)

			// z[2*n:] += As*x[:n]
			x_val = matrix.Times(As, x_n)
			z.AddIndexes(matrix.Indexes(2*n, z.NumElements()), x_val.FloatArray())

			checkpnt.Check("50-f", minor)
			return nil
		}
		return
	}
	
	// matrix(n*[0.0] + n*[1.0])
	c := matrix.FloatZeros(n, 2)
	c.Set(1.0, matrix.ColumnIndexes(c, 1)...)
	matrix.Reshape(c, 2*n, 1)

	h := matrix.FloatZeros(2*n+m+1, 1)
	h.Set(1.0, 2*n)
	// h[2*n+1:] = -b
	h.SetSubMatrix(2*n+1, 0, matrix.Scale(b, -1.0))
	G := &matrixFs{A}

	dims := sets.DSetNew("l", "q", "s")
	dims.Set("l", []int{2*n})
	dims.Set("q", []int{m+1})

	var solopts cvx.SolverOptions
	solopts.MaxIter = maxIter
	solopts.ShowProgress = true
	return cvx.ConeLpCustom(c, G, h, nil, nil, dims, Fkkt, &solopts, nil, nil)
}

func main() {

	var A, b *matrix.FloatMatrix = nil, nil
	m, n := 20, 20

	flag.Parse()
	if len(spPath) > 0 {
		checkpnt.Reset(spPath)
		checkpnt.Activate()
		checkpnt.Verbose(spVerbose)
		checkpnt.Format("%.17f")
	}
	if len(AVal) > 0 {
		A, _ = matrix.FloatParseSpe(AVal)
		if A == nil {
			fmt.Printf("could not parse:\n%s\n", AVal)
			return
		}
	} else {
		A = matrix.FloatNormal(m, n)
	}
	if len(bVal) > 0 {
		b, _ = matrix.FloatParseSpe(bVal)
		if b == nil {
			fmt.Printf("could not parse:\n%s\n", bVal)
			return
		}
	} else {
		b = matrix.FloatNormal(m, 1)
	}

	sol, err := qcl1(A, b)
	if sol != nil {
		r := sol.Result.At("x")[0]
		x := matrix.FloatVector(r.FloatArray()[:A.Cols()])
		r = sol.Result.At("z")[0]
		z := matrix.FloatVector(r.FloatArray()[r.NumElements()-A.Rows():])
		fmt.Printf("x=\n%v\n", x.ToString("%.9f"))
		fmt.Printf("z=\n%v\n", z.ToString("%.9f"))
		check(x, z)
	} else {
		fmt.Printf("status: %v\n", err)
	}
}

// Local Variables:
// tab-width: 4
// End:

File: examples/py/testqcl1.py
# The quadratically constrained 1-norm minimization example of section 8.7
# (Exploiting structure).

import sys
from cvxopt import blas, lapack, solvers, matrix, mul, div, setseed, normal
from math import sqrt
import helpers
import localcones

# helper variables for checkpointing
loopg = 0
loopf = 0

def qcl1(A, b):

    """
    Returns the solution u, z of

        (primal)  minimize    || u ||_1       
                  subject to  || A * u - b ||_2  <= 1

        (dual)    maximize    b^T z - ||z||_2
                  subject to  || A'*z ||_inf <= 1.

    Exploits structure, assuming A is m by n with m >= n. 
    """

    m, n = A.size

    # Solve equivalent cone LP with variables x = [u; v]:
    #
    #     minimize    [0; 1]' * x 
    #     subject to  [ I  -I ] * x <=  [  0 ]   (componentwise)
    #                 [-I  -I ] * x <=  [  0 ]   (componentwise)
    #                 [ 0   0 ] * x <=  [  1 ]   (SOC)
    #                 [-A   0 ]         [ -b ].
    #
    #     maximize    -t + b' * w
    #     subject to  z1 - z2 = A'*w
    #                 z1 + z2 = 1
    #                 z1 >= 0,  z2 >=0,  ||w||_2 <= t.
     
    c = matrix(n*[0.0] + n*[1.0])
    h = matrix( 0.0, (2*n + m + 1, 1))
    h[2*n] = 1.0
    h[2*n+1:] = -b

    def G(x, y, alpha = 1.0, beta = 0.0, trans = 'N'):    
        minor = 0
        if not helpers.sp_minor_empty():
            minor = helpers.sp_minor_top()
        else:
            global loopg
            loopg += 1
            minor = loopg
        helpers.sp_create("00-Gfunc", minor)

        y *= beta
        if trans=='N':
            # y += alpha * G * x 
            y[:n] += alpha * (x[:n] - x[n:2*n]) 
            y[n:2*n] += alpha * (-x[:n] - x[n:2*n]) 
            y[2*n+1:] -= alpha * A*x[:n] 

        else:
            # y += alpha * G'*x 
            y[:n] += alpha * (x[:n] - x[n:2*n] - A.T * x[-m:])  
            y[n:] -= alpha * (x[:n] + x[n:2*n]) 

        helpers.sp_create("10-Gfunc", minor)


    def Fkkt(W): 

        # Returns a function f(x, y, z) that solves
        #
        #     [ 0   G'   ] [ x ] = [ bx ]
        #     [ G  -W'*W ] [ z ]   [ bz ].

        # First factor 
        #
        #     S = G' * W**-1 * W**-T * G
        #       = [0; -A]' * W3^-2 * [0; -A] + 4 * (W1**2 + W2**2)**-1 
        #
        # where
        #
        #     W1 = diag(d1) with d1 = W['d'][:n] = 1 ./ W['di'][:n]  
        #     W2 = diag(d2) with d2 = W['d'][n:] = 1 ./ W['di'][n:]  
        #     W3 = beta * (2*v*v' - J),  W3^-1 = 1/beta * (2*J*v*v'*J - J)  
        #        with beta = W['beta'][0], v = W['v'][0], J = [1, 0; 0, -I].
  
        # As = W3^-1 * [ 0 ; -A ] = 1/beta * ( 2*J*v * v' - I ) * [0; A]
 
        minor = 0
        if not helpers.sp_minor_empty():
            minor = helpers.sp_minor_top()

        beta, v = W['beta'][0], W['v'][0]
        As = 2 * v * (v[1:].T * A)
        As[1:,:] *= -1.0
        As[1:,:] -= A
        As /= beta
      
        # S = As'*As + 4 * (W1**2 + W2**2)**-1
        S = As.T * As 
        d1, d2 = W['d'][:n], W['d'][n:]       
        d = 4.0 * (d1**2 + d2**2)**-1
        S[::n+1] += d
        lapack.potrf(S)

        def f(x, y, z):

            minor = 0
            if not helpers.sp_minor_empty():
                minor = helpers.sp_minor_top()
            else:
                global loopf
                loopf += 1
                minor = loopf
            helpers.sp_create("00-f", minor)
  
            # z := - W**-T * z 
            z[:n] = -div( z[:n], d1 )
            z[n:2*n] = -div( z[n:2*n], d2 )

            z[2*n:] -= 2.0*v*( v[0]*z[2*n] - blas.dot(v[1:], z[2*n+1:]) ) 
            z[2*n+1:] *= -1.0
            z[2*n:] /= beta

              # x := x - G' * W**-1 * z
            x[:n] -= div(z[:n], d1) - div(z[n:2*n], d2) + As.T * z[-(m+1):]
            x[n:] += div(z[:n], d1) + div(z[n:2*n], d2) 
            helpers.sp_create("15-f", minor)

  
            # Solve for x[:n]:
            #
            #    S*x[:n] = x[:n] - (W1**2 - W2**2)(W1**2 + W2**2)^-1 * x[n:]
            
            x[:n] -= mul( div(d1**2 - d2**2, d1**2 + d2**2), x[n:]) 
            helpers.sp_create("25-f", minor)

            lapack.potrs(S, x)
            helpers.sp_create("30-f", minor)
            
            # Solve for x[n:]:
            #
            #    (d1**-2 + d2**-2) * x[n:] = x[n:] + (d1**-2 - d2**-2)*x[:n]
             
            x[n:] += mul( d1**-2 - d2**-2, x[:n])
            helpers.sp_create("35-f", minor)

            x[n:] = div( x[n:], d1**-2 + d2**-2)
            helpers.sp_create("40-f", minor)

            # z := z + W^-T * G*x 
            z[:n] += div( x[:n] - x[n:2*n], d1) 
            helpers.sp_create("44-f", minor)

            z[n:2*n] += div( -x[:n] - x[n:2*n], d2) 
            helpers.sp_create("48-f", minor)

            z[2*n:] += As*x[:n]
            helpers.sp_create("50-f", minor)
  
        return f

    dims = {'l': 2*n, 'q': [m+1], 's': []}
    localcones.options['maxiters'] = 30
    sol = localcones.conelp(c, G, h, dims, kktsolver = Fkkt)
    if sol['status'] == 'optimal':
        return sol['x'][:n],  sol['z'][-m:]
    else:
        return None, None

def rungo(A, b, x, z):
    helpers.run_go_test("../testqcl1", {'x': x, 'z': z, 'A': A, 'b': b})

setseed()
#m, n = 100, 100
m, n = 10, 10
A, b = normal(m,n), normal(m,1)

no_go = False
if len(sys.argv[1:]) > 0 and sys.argv[1] == '-sp':
    helpers.sp_reset("./sp.qcl1")
    helpers.sp_activate()
    no_go = True

x, z = qcl1(A, b)
if x is None:
    print("infeasible")
    x = matrix(0.0, (0, 1))
    z = matrix(0.0, (0, 1))
else:
    print "x\n", helpers.str2(x, "%.9f")
    print "z\n", helpers.str2(z, "%.9f")

if not no_go:
    rungo(A, b, x, z)
else:
    print "A=\"%s\"\n" % helpers.strSpe(A)
    print "b=\"%s\"\n" % helpers.strSpe(b)
